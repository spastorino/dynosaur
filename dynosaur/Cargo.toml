[package]
name = "dynosaur"
version = "0.2.0"
authors = [
    "Santiago Pastorino <spastorino@gmail.com>",
    "Tyler Mandry <tmandry@google.com>",
    "Niko Matsakis <rust@nikomatsakis.com>",
]
categories = ["asynchronous", "no-std"]
keywords = ["async", "trait", "impl"]
description = "Dynamic dispatch for return position impl traits and async in Rust"
documentation = "https://docs.rs/dynosaur"
license = "MIT OR Apache-2.0"
repository = "https://github.com/spastorino/dynosaur"
edition = "2021"
rust-version = "1.75"

[dependencies]
dynosaur_derive = { version = "0.2", path = "../dynosaur_derive" }

[dev-dependencies]
tokio = { workspace = true }
trait-variant = { workspace = true }
ui_test = "0.28"

[[example]]
name = "next"

[[example]]
name = "trait_variant"

[[test]]
name = "ui"
harness = false
