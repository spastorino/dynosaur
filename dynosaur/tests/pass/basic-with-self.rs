use dynosaur::dynosaur;

#[dynosaur(DynMyTrait = dyn(box))]
trait MyTrait {
    type Item;
    async fn foo(&self) -> Self::Item;
}

fn main() {}
