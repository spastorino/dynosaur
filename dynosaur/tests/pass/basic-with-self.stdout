#![feature(prelude_import)]
#[prelude_import]
use std::prelude::rust_2021::*;
#[macro_use]
extern crate std;
use dynosaur::dynosaur;

trait MyTrait {
    type Item;
    async fn foo(&self)
    -> Self::Item;
}
mod __dynosaur_macro_dynmytrait {
    use super::*;
    trait ErasedMyTrait {
        type Item;
        fn foo<'life0, 'dynosaur>(&'life0 self)
        ->
            ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
            Self::Item> + 'dynosaur>>
        where
        'life0: 'dynosaur,
        Self: 'dynosaur;
    }
    impl<DYNOSAUR: MyTrait> ErasedMyTrait for DYNOSAUR {
        type Item = <Self as MyTrait>::Item;
        fn foo<'life0, 'dynosaur>(&'life0 self)
            ->
                ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                Self::Item> + 'dynosaur>> where 'life0: 'dynosaur,
            Self: 'dynosaur {
            Box::pin(<Self as MyTrait>::foo(self))
        }
    }
    #[repr(transparent)]
    pub struct DynMyTrait<'dynosaur_struct, Item> {
        ptr: dyn ErasedMyTrait<Item = Item> + 'dynosaur_struct,
    }
    impl<'dynosaur_struct, Item> MyTrait for
        DynMyTrait<'dynosaur_struct, Item> {
        type Item = Item;
        fn foo(&self) -> impl ::core::future::Future<Output = Self::Item> {
            let ret:
                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                    Self::Item> + '_>> = self.ptr.foo();
            let ret:
                    ::core::pin::Pin<Box<dyn ::core::future::Future<Output =
                    Self::Item> + 'static>> =
                unsafe { ::core::mem::transmute(ret) };
            ret
        }
    }
    impl<'dynosaur_struct, Item> DynMyTrait<'dynosaur_struct, Item> {
        pub fn new_box(value: impl MyTrait<Item = Item> + 'dynosaur_struct)
            -> Box<DynMyTrait<'dynosaur_struct, Item>> {
            let value = Box::new(value);
            let value:
                    Box<dyn ErasedMyTrait<Item = Item> + 'dynosaur_struct> =
                value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_arc(value: impl MyTrait<Item = Item> + 'dynosaur_struct)
            -> std::sync::Arc<DynMyTrait<'dynosaur_struct, Item>> {
            let value = std::sync::Arc::new(value);
            let value:
                    std::sync::Arc<dyn ErasedMyTrait<Item = Item> +
                    'dynosaur_struct> = value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub fn new_rc(value: impl MyTrait<Item = Item> + 'dynosaur_struct)
            -> std::rc::Rc<DynMyTrait<'dynosaur_struct, Item>> {
            let value = std::rc::Rc::new(value);
            let value:
                    std::rc::Rc<dyn ErasedMyTrait<Item = Item> +
                    'dynosaur_struct> = value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub const fn from_ref(value:
                &(impl MyTrait<Item = Item> + 'dynosaur_struct))
            -> &DynMyTrait<'dynosaur_struct, Item> {
            let value: &(dyn ErasedMyTrait<Item = Item> + 'dynosaur_struct) =
                &*value;
            unsafe { ::core::mem::transmute(value) }
        }
        pub const fn from_mut(value:
                &mut (impl MyTrait<Item = Item> + 'dynosaur_struct))
            -> &mut DynMyTrait<'dynosaur_struct, Item> {
            let value:
                    &mut (dyn ErasedMyTrait<Item = Item> + 'dynosaur_struct) =
                &mut *value;
            unsafe { ::core::mem::transmute(value) }
        }
    }
    impl<DYNOSAUR: MyTrait> MyTrait for &DYNOSAUR where DYNOSAUR: ?Sized {
        type Item = <DYNOSAUR as MyTrait>::Item;
        fn foo(&self) -> impl ::core::future::Future<Output = Self::Item> {
            <DYNOSAUR as MyTrait>::foo(self)
        }
    }
    impl<DYNOSAUR: MyTrait> MyTrait for &mut DYNOSAUR where DYNOSAUR: ?Sized {
        type Item = <DYNOSAUR as MyTrait>::Item;
        fn foo(&self) -> impl ::core::future::Future<Output = Self::Item> {
            <DYNOSAUR as MyTrait>::foo(self)
        }
    }
    impl<DYNOSAUR: MyTrait> MyTrait for Box<DYNOSAUR> where DYNOSAUR: ?Sized {
        type Item = <DYNOSAUR as MyTrait>::Item;
        fn foo(&self) -> impl ::core::future::Future<Output = Self::Item> {
            <DYNOSAUR as MyTrait>::foo(self)
        }
    }
}
use __dynosaur_macro_dynmytrait::DynMyTrait;

fn main() {}
